node: documentation for add question, answer question, and vote code

wooldridge opened the pull request.

wooldridge committed:
node: documentation for add, answer, and vote code

Questions:

contributor/index.js, line 30
I don't see patchVoteCount() called anywhere in /lib. Can we delete?

qnaDoc/patch.js, line 132
The spec param is passed in with an empty comments property. Is this needed?

qnaDoc/index.js, line 31
The spec param can specify a single document or multiple (as coded). Where is an example of that method handling multiple documents? I don't see any examples. In such a case, is spec an array of objects (similar to the object for the singular case)?

wooldridge commented on appserver/node-express/lib/db-client/contributor/index.js:
```diff
@@ -24,6 +27,7 @@ funcs.patchReputation = function (txid, id, repChange) {
   ).result();
 };
 
+// TODO I don't see patchVoteCount() called anywhere in /lib. Can we delete?
```
I don't see patchVoteCount() called anywhere in /lib. Can we delete?

laurelnaiad replied:
This may be a bug, I don't recall where the spec landed, let's double-check that we aren't supposed to be calling it before we delete it.

wooldridge replied:
Method is not used anywhere and its functionality is handled in votePatch() in patch.js. I will remove.

wooldridge commented on appserver/node-express/lib/db-client/qnaDoc/index.js:
```diff
@@ -22,6 +22,18 @@ funcs.search = function (spec) {
+ * @param  {String} txid The transaction ID.
+ * @param  {Object|Array} spec An object with a document ID property. Example:
+ *   {"id":"b6ff5e47-2d08-4c20-9a16-4998fbb1bc10"}
+ * For multiple documents, an array of objects (TODO Is this correct? Where is
```
The spec param can specify a single document or multiple (as coded). Where is an example of that method handling multiple documents? I don't see any examples. In such a case, is spec an array of objects (similar to the object for the singular case)?

laurelnaiad replied:
There are a few reasons it is this way.
1. I think we may hit the multiple document scenario in configuring the database and/or testing
2. In many different promise chains, we need to know how many results we got, and being able to say in the error message that we expected 1 and got, for instance, 3, is of value, so that function (cardinatity check) is built into the getUnique generalized function. That's why in this particular function we don't go the distance to make sure that there's only one -- it's generalized elsewhere (in part because of #1).

Anyway, it could have gone a few different directions, the design as is leaves the code more generalized so that as features are added which require getting more than one document, this function can remain in service (especially if its method of turning the spec into criteria for the client api is improved,which it should be).

wooldridge replied:
OK.

wooldridge commented on appserver/node-express/lib/db-client/qnaDoc/patch.js:
```diff
@@ -66,6 +116,25 @@ module.exports = function (txid, spec) {
+   * @param  {String} questionId The question ID.
+   * @param  {Object} spec Answer content with "comments" and "text" properties.
+   * The comments property is an empty array (TODO Why? Remove?). Example:
+   *   {"comments":[],
```
The spec param is passed in with an empty comments property. Is this needed?

laurelnaiad replied:
Are you saying comments need to be supplied, or we're calling this function with comments that need not necessarily be supplied? I believe what we end up storing is set to match the Java code's behavior. If we don't need to add empty commebnts, then we shouldn't. If we **do** need empty comments, we should add them **here** as that requirement is really a db-client concern (probably related to the indexing) rather than an application-level necessity).

wooldridge replied:
Thanks. I see that the spec (and thus the comments) comes in (from the routing) from the req object, and the empty comments is not of db-client concern, so not necessary to include specifics here.

laurelnaiad commented on appserver/node-express/lib/db-client/contributor/index.js:
```diff
@@ -6,11 +6,14 @@ var util = libRequire('db-client/util');
+ *
+ * @param  {String} txid The transaction ID.
+ * @param  {String} id The contributor ID.
+ * @param  {Number} repChange Increment (1) or decrement (-1) value.
```
+10 would work as would -3. Let's document this function as abstractly as its written, where repChange is expected to be a signed integer. Other functions are what worry about setting the value. This one intentionally just does addition of a (positive or negative) number.

wooldridge replied:
Right, thanks for catching this.

wooldridge committed:
node: remove patchVoteCount() in contributor/index.js since not used.

Nothing in /lib is using this method. This function is handled by votePatch() in patch.js.

wooldridge committed:
node: vote methods can take arbitrary increment/decrement integers, document as such

wooldridge committed:
node: update comments in patchQuestionAddAnswer().

laurelnaiad commented on appserver/node-express/lib/db-client/qnaDoc/patch.js:
```diff
@@ -66,6 +116,24 @@ module.exports = function (txid, spec) {
+   *    "displayName":"JoeUser"}
+   * @param  {String} questionId The question ID.
+   * @param  {Object} spec The answer content from the request. Example:
+   *   {"comments":[],
```
Someone seems to have deleted a discussion about comments... the concern I have is that if an empty comments array needs to be included in the document when it is loaded, it is very much this function's concern to provide that, and not the router, as it's a storage-level problem, not an app-logic problem. Let's check whether or not the empty array is required, and if so, this function should assure that it is present before attempting to store the document.

wooldridge replied:
I figured out why the empty comments array is there when it seems like it doesn't need to be. I'd forgotten how it all works in the browser tier. We create a "draft" answer for our answer form on the QnADoc page. That draft answer has an empty comments array (as part of the standard ssAnswer creation) even though there's nothing that would be put in the comment array from the new answer form. (The new answer form has fields for just the title and the body text.) 

We can delete that empty comments array for the draft answer (after its creation) or we can leave the array there without affecting how the app works to the end user. But considering that Dave's students will be working with the new answer functionality, removing the comments array would make it slightly less confusing IMO. And removing that comments array means we wouldn't have to explain (in our code comments) why that empty array is there. @laurelnaiad any thoughts?

laurelnaiad replied:
Actually I've at least partially reworked the "draft" stuff in the browser. If I left it such that it's posting empty arrays, then so be it. 

I don't know I fully understand how this array came to be of interest to the dbClient function, so what to do about it kind of depends on who is putting it there and why it needs to be there (if at all). 
- If it's coming from the browser into the middle tier, then by all means the middle tier should dutifully store it, since that property being an empty array is entirely valid. 
- If the middle tier is generating it "for no particular reason," and both the Java and Node tiers work just fine without it being there at all, then it probably shouldn't be generating it all on its own.
- If the middle tier is generating it because it's necessary (for the indexing code, for instance), then it should be generated by the dbClient code and not the router/app logic code, as it would be a "phyiscal data need" rather than a logical need to express the lack of comments.

Does that make sense?

laurelnaiad replied:
Followup -- just to be clear -- if the browser sends an empty array, we should leave it there (in the middle tier).... it's totally valid for the browser to claim that there are 0 comments on something if there aren't any. :smile:

wooldridge replied:
Great, thanks for the clarification. The comments are coming in from the browser (your first bullet above).  I will have the middle tier dutifully store it (which means making no changes in the middle tier; it's doing what it needs to do now). 

Now that I revisited how our "draft" answer works, I understand why that empty comments array is there.

wooldridge committed:
node: remove unneeded comments array in draft answer

wooldridge committed:
revert the last commit, don't delete the comment array in the answer coming from the browser.

wooldridge committed:
static site update of last commit
